### PR TITLE
feat(rule): heading-whitespace-only

### PR DESCRIPTION
**Pull request by straker**
New rule for evaluating headings that only contain certain space separator unicode characters or `<br>` and `<wbr>` elements.

Closes: #1419

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [ ] Follows the commit message policy, appropriate for next version
- [ ] Has documentation updated, a DU ticket, or requires no documentation change
- [ ] Includes new tests, or was unnecessary
- [ ] Code is reviewed for security by: << Name here >>

### CODE REVIEW

File: lib/checks/shared/heading-whitespace-only.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+	"id": "heading-whitespace-only",
+	"evaluate": "heading-whitespace-only.js",
+	"metadata": {
+		"impact": "minor",
+		"messages": {
+			"pass": "Element is empty or does not have whitespace",
+			"fail": "Element does not have text that is visible to screen readers and contains only whitespace characters",
+			"incomplete": ""
```
Comment (jkodu) [inline lib/checks/shared/heading-whitespace-only.json]: nitpick: remove `incomplete`

File: lib/rules/heading-whitespace-only.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+	"id": "heading-whitespace-only",
+	"selector": "h1, h2, h3, h4, h5, h6, [role=heading]",
+	"matches": "heading-matches.js",
+	"tags": ["wcag2a", "wcag131"],
+	"metadata": {
+		"description": "Ensure that heading elements do not have Unicode separator characters, `<br>`, or `<wbr>` elements as their only content",
+		"help": "Headings should contain text and not just whitespace or word break elements."
```
Comment (jkodu) [inline lib/rules/heading-whitespace-only.json]: editorial: heading elements should ...

File: lib/rules/heading-whitespace-only.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	},
+	"preload": false,
```
Comment (jkodu) [inline lib/rules/heading-whitespace-only.json]: you should remove `preload` flag

File: lib/checks/shared/heading-whitespace-only.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+if (axe.commons.text.accessibleTextVirtual(virtualNode)) {
+	return true;
+}
+
+const spaceSeparatorCharCodes = [
```
Comment (jkodu) [inline lib/checks/shared/heading-whitespace-only.js]: may be worth using something like this - https://www.npmjs.com/package/is-whitespace-character?

File: lib/checks/shared/heading-whitespace-only.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	160, // non-breaking space
+	8194, // en space
+	8195, // em space
+	8201 // thin space
+];
+
+let hasBreakingSpaceElement = node.querySelectorAll('br, wbr').length > 0;
```
Comment (jkodu) [inline lib/checks/shared/heading-whitespace-only.js]: could return here immediately, and do the `hasSpaceCharacter` check after.

File: lib/checks/shared/heading-whitespace-only.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+let hasSpaceSeparator =
+	node.textContent
```
Comment (jkodu) [inline lib/checks/shared/heading-whitespace-only.js]: @WilcoFiers - is there a better util method we can use instead of textContent.

@straker - does the `accessibleTextVirtual` trim out spaces? Perhaps worth passing an argument into it to not trim spaces and use that value from above here?

I remember using `textContent` in the past & it was asked to be changed (can't remember why).

File: test/checks/shared/heading-whitespace-only.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
+describe('heading-whitespace-only', function() {
+	'use strict';
```
Comment (jkodu) [inline test/checks/shared/heading-whitespace-only.js]: some more edge cases:
- what happens when heading is given accessible name via `aria-*`? and they end up being space characters?
- what happens if heading element has shadowDOM that then exposes a non space character as heading content?

File: lib/checks/shared/heading-whitespace-only.js (hunk 4 above)
Comment (straker) [reply lib/checks/shared/heading-whitespace-only.js]: I had asked Wilco, and he wanted to just start with the 4 failing cases listed in https://github.com/dequelabs/axe-core/issues/1419

File: lib/checks/shared/heading-whitespace-only.js (hunk 6 above)
Comment (straker) [reply lib/checks/shared/heading-whitespace-only.js]: Ya,` accessibleTextVirtual` will return a 0 length string for all space separator characters

File: test/checks/shared/heading-whitespace-only.js (hunk 7 above)
Comment (straker) [reply test/checks/shared/heading-whitespace-only.js]: I'll have to figure out how to handle those. Thanks for bringing them up.

Comment (dylanb) [reply test/checks/shared/heading-whitespace-only.js]: Definitely need to do some screen reader testing on the aria-label combinations

File: lib/commons/text/accessible-text-virtual.js [hunk 8]
```diff
@@ -56,7 +56,7 @@ text.accessibleTextVirtual = function accessibleTextVirtual(
 		return step(virtualNode, context);
 	}, '');
 
-	if (context.startNode === virtualNode) {
+	if (context.sanitize && context.startNode === virtualNode) {
```
Comment (WilcoFiers) [inline lib/commons/text/accessible-text-virtual.js]: Strong preference for using `context.unsanitized`. That way all truthy values do one thing, and all falsey values do another, as opposed to now where undefined behaves as though it's truthy.

File: lib/commons/text/accessible-text-virtual.js [hunk 9]
```diff
@@ -128,6 +128,9 @@ function prepareContext(virtualNode, context) {
 			...context
 		};
 	}
+	if (typeof context.sanitize === 'undefined') {
+		context.sanitize = true;
```
Comment (WilcoFiers) [inline lib/commons/text/accessible-text-virtual.js]: Better to treat the context as readonly if you can.

File: lib/checks/shared/heading-whitespace-only.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+	8194, // en space
+	8195, // em space
+	8201 // thin space
+];
+
+const hasSpaceSeparator =
+	text
+		.split('')
+		.map(char => char.charCodeAt(0))
+		.filter(charCode => spaceSeparatorCharCodes.includes(charCode)).length > 0;
```
Comment (WilcoFiers) [inline lib/checks/shared/heading-whitespace-only.js]: This looks wrong to me. `<h3>Foo&160;bar</h3>` shouldn't fail this rule. I think what this should be is this:

```js
const textCharCodes = text.split('').map(char => char.charCodeAt(0))
const hasOnlySpaceSeparator = textCharCodes.every(charCode => {
  return spaceSeparatorCharCodes.includes(charCode)
})
```

File: test/checks/shared/heading-whitespace-only.js [hunk 11]
```diff
@@ -0,0 +1,214 @@
+describe('heading-whitespace-only', function() {
```
Comment (WilcoFiers) [inline test/checks/shared/heading-whitespace-only.js]: You're not testing combinations of things, multiple chars that should fail, chars that should fail together with chars that pass, chars mixed with break elements. Most of these tests can be integration tests if you move the chars and breaking space elements into an option object.

File: lib/checks/shared/heading-whitespace-only.js (hunk 10 above)
Comment (straker) [reply lib/checks/shared/heading-whitespace-only.js]: Should `<h3>   &160;   </h3>` also fail the rule?

**Comment by jkodu** [comment]
Upon discussing about the status of the rule in [ACT-Rules](https://github.com/act-rules/act-rules.github.io/pull/400), which is going to go through some changes with `applicability` and `expectation`, it would be good to pause this rule for the time being.

I will re-open the same, when more clarity/ consensus of the above mentioned rule.
